substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/meta/AnalysisType.java

```diff
@@ -542,33 +544,32 @@ private static void updateFlow(BigBang bb, TypeFlow<?> flow, TypeState newState,
         }
     }
 
-    public void registerAsInHeap() {
-        if (!isInHeap) {
-            /* Races are not a problem because every thread is going to do the same steps. */
-            isInHeap = true;
-
+    public boolean registerAsInHeap() {
+        boolean firstAttempt = atomicMark(isInHeap);
+        if (firstAttempt) {
             assert isArray() || (isInstanceClass() && !Modifier.isAbstract(getModifiers())) : this;
             universe.hostVM.checkForbidden(this, UsageKind.InHeap);
         }
         registerAsReachable();
```

Comment:
For cleanliness, we should move the `registerAsReachable` before the `atomicMark`. Again, now that you are introducing some sort of atomicity guarantees, there should never be a time window where a type is marked as allocated but not yet as reachable.

Reply:
Done.